File: windows/tests/DungeonSoundboard.Core.Tests/TrackTests.cs
using System.Text.Json;
using DungeonSoundboard.Core.Models;
using DungeonSoundboard.Core.Serialization;
using Xunit;

namespace DungeonSoundboard.Core.Tests;

public sealed class TrackTests
{
    [Fact]
    public void DecodingLegacyTrackDefaultsVolumeMultiplierToOne()
    {
        const string json = """
        {
          "id": "00000000-0000-0000-0000-000000000001",
          "title": "Rain",
          "path": "C:\\audio\\rain.mp3",
          "role": "music"
        }
        """;

        var track = JsonSerializer.Deserialize<Track>(json, JsonDefaults.Options);

        Assert.NotNull(track);
        Assert.Equal(1.0, track.VolumeMultiplier);
    }

    [Fact]
    public void VolumeMultiplierIsClamped()
    {
        var quiet = new Track("Quiet", "C:\\audio\\quiet.mp3", volumeMultiplier: -2);
        Assert.Equal(0, quiet.VolumeMultiplier);

        quiet.VolumeMultiplier = 3;
        Assert.Equal(2, quiet.VolumeMultiplier);

        var invalid = new Track("Invalid", "C:\\audio\\invalid.mp3", volumeMultiplier: double.NaN);
        Assert.Equal(1, invalid.VolumeMultiplier);
    }

    [Fact]
    public void NormalizedTitleTrimsAndFallsBack()
    {
        Assert.Equal("Tavern Rain", Track.NormalizedTitle("  Tavern Rain  ", "Rain"));
        Assert.Equal("Rain", Track.NormalizedTitle("   ", "Rain"));
    }

    [Fact]
    public void OutputVolumeUsesMasterTrackMultiplierAndDucking()
    {
        Assert.True(Math.Abs(Track.OutputVolume(0.5, 1.5, 0.5) - 0.375) < 0.0001);
        Assert.True(Math.Abs(Track.OutputVolume(1.0, 2.0) - 1.0) < 0.0001);
        Assert.True(Math.Abs(Track.OutputVolume(0.2, 2.0) - 0.4) < 0.0001);
    }

    [Fact]
    public void MoveTrackReordersIds()
    {
        var firstId = Guid.NewGuid();
        var secondId = Guid.NewGuid();
        var thirdId = Guid.NewGuid();
        var tracks = new List<Track>
        {
            new("First", "C:\\audio\\first.mp3", id: firstId),
            new("Second", "C:\\audio\\second.mp3", id: secondId),
            new("Third", "C:\\audio\\third.mp3", id: thirdId)
        };

        Assert.True(Track.MoveTrack(tracks, firstId, thirdId));
        Assert.Equal([secondId, thirdId, firstId], tracks.Select(track => track.Id).ToArray());

        Assert.True(Track.MoveTrack(tracks, firstId, secondId));
        Assert.Equal([firstId, secondId, thirdId], tracks.Select(track => track.Id).ToArray());
    }
}

File: Tests/DungeonSoundboardCoreTests/TrackTests.swift
import XCTest
@testable import DungeonSoundboardCore

final class TrackTests: XCTestCase {
    func testDecodingLegacyTrackDefaultsVolumeMultiplierToOne() throws {
        let json = """
        {
          "id": "00000000-0000-0000-0000-000000000001",
          "title": "Rain",
          "path": "/tmp/rain.mp3",
          "role": "music"
        }
        """
        let track = try JSONDecoder().decode(Track.self, from: Data(json.utf8))

        XCTAssertEqual(track.volumeMultiplier, 1.0)
    }

    func testVolumeMultiplierIsClamped() {
        var quiet = Track(title: "Quiet", path: "/tmp/quiet.mp3", volumeMultiplier: -2)
        XCTAssertEqual(quiet.volumeMultiplier, 0)

        quiet.volumeMultiplier = 3
        XCTAssertEqual(quiet.volumeMultiplier, 2)

        let invalid = Track(title: "Invalid", path: "/tmp/invalid.mp3", volumeMultiplier: .nan)
        XCTAssertEqual(invalid.volumeMultiplier, 1)
    }

    func testNormalizedTitleTrimsAndFallsBack() {
        XCTAssertEqual(Track.normalizedTitle("  Tavern Rain  ", fallback: "Rain"), "Tavern Rain")
        XCTAssertEqual(Track.normalizedTitle("   ", fallback: "Rain"), "Rain")
    }

    func testOutputVolumeUsesMasterTrackMultiplierAndDucking() {
        XCTAssertEqual(
            Track.outputVolume(masterVolume: 0.5, trackMultiplier: 1.5, duckingMultiplier: 0.5),
            0.375,
            accuracy: 0.0001
        )
        XCTAssertEqual(
            Track.outputVolume(masterVolume: 1.0, trackMultiplier: 2.0),
            1.0,
            accuracy: 0.0001
        )
        XCTAssertEqual(
            Track.outputVolume(masterVolume: 0.2, trackMultiplier: 2.0),
            0.4,
            accuracy: 0.0001
        )
    }

    func testMoveTrackReordersIDs() {
        let firstID = UUID()
        let secondID = UUID()
        let thirdID = UUID()
        var tracks = [
            Track(id: firstID, title: "First", path: "/tmp/first.mp3"),
            Track(id: secondID, title: "Second", path: "/tmp/second.mp3"),
            Track(id: thirdID, title: "Third", path: "/tmp/third.mp3")
        ]

        XCTAssertTrue(Track.moveTrack(in: &tracks, draggedID: firstID, to: thirdID))
        XCTAssertEqual(tracks.map(\.id), [secondID, thirdID, firstID])

        XCTAssertTrue(Track.moveTrack(in: &tracks, draggedID: firstID, to: secondID))
        XCTAssertEqual(tracks.map(\.id), [firstID, secondID, thirdID])
    }
}
